sagemaker/pipeline/pipeline_parameters.py

```diff
@@ -0,0 +1,764 @@
+                f"got job sequence: {self.jobs_to_run}"
+            )
+
+        # Fix trailing slash for S3 input paths, otherwise can get
+        # ClientError: Failed to invoke sagemaker:CreateProcessingJob. Error Details: null.
+        if self.input_data_s3.endswith("/"):
+            self.input_data_s3 = self.input_data_s3[:-1]
+
+
+@dataclass
+class GraphConstructionConfig:
+    """Configuration for the graph construction step"""
+
+    config_filename: str
+    graph_construction_args: str
+
+
+@dataclass
+class PartitionConfig:
+    """Configuration for the partition step"""
+
+    partition_algorithm: str
+    input_json_filename: str
+    output_json_filename: str
+
+
+@dataclass
+class TrainingConfig:
+    """Configuration for the training step"""
+
+    model_output_path: str
+    train_inference_task: str
+    train_yaml_file: str
+    num_trainers: int
+    use_graphbolt_str: str
+
+    def __post_init__(self):
+        self.use_graphbolt_str = self.use_graphbolt_str.lower()
+
+
+@dataclass
+class InferenceConfig:
+    """Configuration for the inference step"""
+
+    save_embeddings: bool
+    save_predictions: bool
+    inference_model_snapshot: str
+    inference_yaml_file: str
+
+
+@dataclass
+class ScriptPaths:
+    """Entry point script locations"""
+
+    dist_part_script: str
+    gb_convert_script: str
+    train_script: str
+    inference_script: str
+    gconstruct_script: str
+    gsprocessing_script: str
+
+
+@dataclass()
+class PipelineArgs:
+    """Wrapper class for all pipeline configurations.
+
+    Parameters
+    ----------
+    aws_config : AWSConfig
+        AWS configuration settings
+    graph_construction_config : GraphConstructionConfig
+        Graph construction configuration
+    instance_config : InstanceConfig
+        Instance configuration settings
+    task_config : TaskConfig
+        Task-level configuration settings
+    partition_config : PartitionConfig
+        Partition configuration settings
+    training_config : TrainingConfig
+        Training configuration settings
+    inference_config : InferenceConfig
+        Inference configuration settings
+    script_paths : ScriptPaths
+        Paths to SageMaker entry point scripts
+    step_cache_expiration : str
+        Cache expiration for pipeline steps
+    update : bool
+        Whether to update existing pipeline or create a new one
+    """
+
+    aws_config: AWSConfig
+    graph_construction_config: GraphConstructionConfig
+    instance_config: InstanceConfig
+    task_config: TaskConfig
+    partition_config: PartitionConfig
+    training_config: TrainingConfig
+    inference_config: InferenceConfig
+    script_paths: ScriptPaths
+    step_cache_expiration: str
+    update: bool
+
+    def __eq__(self, other):
+        if not isinstance(other, PipelineArgs):
+            return False
+        return asdict(self) == asdict(other)
+
+    def __hash__(self):
```

Comment:
Why we want to overwrite `__hash__`

Reply:
Just to ensure we control the implementation. Creation of `__hash__` is a bit convoluted for data classes so I want to be explicit.

From https://docs.python.org/3/library/dataclasses.html

> If eq and frozen are both true, by default @dataclass will generate a __hash__() method for you. If eq is true and frozen is false, __hash__() will be set to None, marking it unhashable (which it is, since it is mutable). If eq is false, __hash__() will be left untouched meaning the __hash__() method of the superclass will be used (if the superclass is [object](https://docs.python.org/3/library/functions.html#object), this means it will fall back to id-based hashing).